[Compound] Integrate compound tokens

jmartinesp opened the pull request:
## Changes

- Added Compound tokens for both colors and typography inside a new `:libraries:theme` module, moved some other classes related to color tokens there.
- Mapped our `ElementTheme` colors and typography to these new tokens.
- Added extra previews of components so the Compound team can test them in Showkase.

## TODO

These are tasks that are out of scope for this PR but will need to eventually be done:

- Slowly migrate screens and components to use Compound tokens for colors and typography.
- Replace tokens in `ElementColors` with Compound colors (or remove them and use the Compound tokens directly).
- Remove `ElementTextStyles` and use `MaterialTheme.typography` tokens instead.

julioromano commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/ElementTheme.kt:
```diff
@@ -46,20 +51,26 @@ object ElementTheme {
 
 /* Global variables (application level) */
 val LocalColors = staticCompositionLocalOf { elementColorsLight() }
+val LocalCompoundColors = staticCompositionLocalOf { compoundColorsLight }
```
It is more ergonomic if we surface this API as an extension over `MaterialTheme` and hide the composition locals (i.e. make them internal/private implementation details).

Access to these colors could then be made with code such as: `MaterialTheme.compoundColors`

draft of impl:
```
val MaterialTheme.compoundColors
    @Composable
    get() = LocalCompoundColors.current
```

julioromano commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/compound/LightDesignTokens.kt:
```diff
@@ -0,0 +1,333 @@
+
+import androidx.compose.ui.graphics.Color
+
+object LightDesignTokens {
```
Consider making these objects `internal` so that developers don't use them.

To better enforce this separation I'd extract a `theme` module from the `designsystem` module and move all theme-related code in there.
This way we are sure that everything marked as `internal` in the `theme` module can't be used by the rest of the code.
`designsystem` can import the `theme` module as  an `api` depdendency so that we don't have to import it everywhere but we'll get it as part of the `designsystem` dependency.

julioromano commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/components/TimePicker.kt:
```diff
@@ -0,0 +1,94 @@
+@Preview(widthDp = 600, group = PreviewGroup.DateTimePickers)
+@Composable
+internal fun TimePickerHorizontalPreview() {
+    ElementThemedPreview {
```
Please move all the "fake" components (the ones that developers shouldn't use but are only needed by the designers to check renderings) in a different subpacakge (perhaps named `internal` or `dontuse` and place a `README.md` file in it which explain why these components should not be used and what are they here for.)

julioromano commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/compound/LightDesignTokens.kt:
```diff
@@ -0,0 +1,333 @@
+package io.element.android.libraries.designsystem.theme.compound
+
+import androidx.compose.ui.graphics.Color
+
```
Is it easy enought to add a comment here that explains these are autogenerated files?

julioromano reviewed (commented):
A few comments as anticipated, let me know if they are not clear enough!

bmarty commented on tests/uitests/src/test/snapshots/images/io.element.android.tests.uitests_ScreenshotTest_preview_tests[io.element.android.libraries.designsystem.theme.components_null_DateTimepickers_DatePickerPreviewLight_0_null,NEXUS_5,1.0,en].png:
The screenshot is containing the current date, so it will fail every time the day is changing, and this happen a lot :).

Can we force a specific date for the preview?

bmarty commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/components/DatePicker.kt:
```diff
@@ -0,0 +1,71 @@
+            buttons = { /*TODO*/ },
+            icon = { /*TODO*/ },
+            title = { /*TODO*/ },
+            text = { DatePicker(state = rememberDatePickerState(), showModeToggle = true) },
```
See my comment on the screenshot, to avoid having the current date recorded.

bmarty commented on libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/theme/components/DatePicker.kt:
```diff
@@ -0,0 +1,71 @@
+            buttons = { /*TODO*/ },
+            icon = { /*TODO*/ },
+            title = { /*TODO*/ },
+            text = { DatePicker(state = rememberDatePickerState(), showModeToggle = true) },
```
Same!

jmartinesp commented:
I think all the issues raised should be fixed with the latest changes:

- I created a `:libraries:theme` module and added everything related to tokens there, then included it with `api` in `designsystem`.
- This new module has a `.compound.gen` package where files generated from Compound are stored, and it's displayed with a different color in the IDE. Also, files inside it should contain a warning now. These files can be replaced by just drag & dropping new ones from the [compound-design-tokens repo](https://github.com/vector-im/compound-design-tokens) (as soon as the changes I made for that are merged there).
- I moved the preview only files inside a `previews` package in `designsystem`, and changed their names so it's clear they only contain previews (also fixed the issues with the Date/Time pickers previews).
- Added some handles in `ElementTheme` for `elementColors`, `compoundColors`, `materialColors` and `typography`.
- Created a README file in `:libraries:theme` with info about the module and its tokens.

bmarty reviewed (approved):
Let's merge this amazing work!
